src/python/pants/option/option_types_test.py

```diff
@@ -0,0 +1,279 @@
+
+        ],
+        # @TODO: These should be in definition order
+        any_order=True,
+    )
+
+    my_subsystem = MySubsystem()
+    assert my_subsystem.str_prop == ""
+    assert my_subsystem.int_prop == 0
+    assert my_subsystem.float_prop == 1.0
+    assert not my_subsystem.bool_prop
+    assert my_subsystem.enum_prop == MyEnum.Val1
+    assert my_subsystem.target_prop == ""
+    assert my_subsystem.dir_prop == "."
+    assert my_subsystem.file_prop == "."
+    assert my_subsystem.shellstr_prop == ""
+    assert my_subsystem.memorysize_prop == 22
+    assert my_subsystem.str_list_prop == ("str1", "str2")
+    assert my_subsystem.int_list_prop == (1, 2)
+    assert my_subsystem.float_list_prop == (1.0, 2.0)
+    assert my_subsystem.bool_list_prop == (False, True)
+    assert my_subsystem.enum_list_prop == (MyEnum.Val2, MyEnum.Val1)
+    assert my_subsystem.target_list_prop == ("str1", "# Once we test the types a better way, these should probably be set to values
+    assert my_subsystem.shellstr_list_prop == ("str1", "str2")
+    assert my_subsystem.memorysize_list_prop == (33, 88)
+    assert my_subsystem.defaultless_str_list_prop == ("str1", "str2")
+    assert my_subsystem.defaultless_int_list_prop == (1, 2)
+    assert my_subsystem.defaultless_float_list_prop == (1.0, 2.0)
+    assert my_subsystem.defaultless_bool_list_prop == (False, True)
+    assert my_subsystem.defaultless_enum_list_prop == (MyEnum.Val2, MyEnum.Val1)
+    assert my_subsystem.defaultless_target_list_prop == ("str1", "str2")
+    assert my_subsystem.defaultless_dir_list_prop == ("str1", "str2")
+    assert my_subsystem.defaultless_file_list_prop == ("str1", "str2")
+    assert my_subsystem.defaultless_shellstr_list_prop == ("str1", "str2")
+    assert my_subsystem.defaultless_memorysize_list_prop == (33, 88)
+
+    # These not only assert that we got the right value out of the `options` object, but also
+    # (indirectly) test that our property type is actually `Optional[T]` and not accidentally `T`.
+    # It does so by relying on two things:
+    #   - If the type was `T`, these assertions would always trigger
+    #   - Because there's code following these asserts, mypy would error that the code is
+    #     unreachable if we got this wrong.
+    assert my_subsystem.optional_str_prop is None
+    assert my_subsystem.optional_int_prop is None
+    assert my_subsystem.optional_float_prop is None
+    assert my_subsystem.optional_bool_prop is None
+    assert my_subsystem.optional_enum_prop is None
+    assert my_subsystem.optional_target_prop is None
+    assert my_subsystem.optional_dir_prop is None
+    assert my_subsystem.optional_file_prop is None
+    assert my_subsystem.optional_shellstr_prop is None
+    assert my_subsystem.optional_memorysize_prop is None
+
+    # This "tests" (through mypy) that the property types are what we expect
+    # Ideally, we'd test these with pytest-mypy-plugins (or similar)
```

Comment:
Before this submits I'd like to try `pytest-mypy-plugins` in a separate PR and use it here.

Reply:
Started Slack: https://pantsbuild.slack.com/archives/C0D7TNJHL/p1643838778597069

Reply:
Actually, lets do it in a followup